_posts/2020-02-17-constructor-injection.md

```diff
@@ -0,0 +1,189 @@
+```java
+@Component
+public class Cake {
+
+	private Logger LOGGER = LoggerFactory.getLogger(Cake.class);
+	
+	private Flavor flavor;
+	
+	@Autowired
+	private Topping toppings;
+	
+	public Cake() {
+		LOGGER.info("Flavor from setter Injection : " + this.flavor);
+	}
+	
+	@Autowired
+	public void setFlavor(Flavor flavor) {
+		LOGGER.info("Initialising flavor object using setter injection");
+		this.flavor = flavor;
+	}
+	...
+}
+
+```
+
+# Why should I use constructor injection?
+
+Now that we have seen the different types of injection, let us go through some of the advantages of using constructor injection.
+
+## All required dependencies are available at initialization time
+
+Every class has a default constructor unless we explicitly create a constructor and its object is created by calling respective constructor. If we specify all the required dependencies in the constructor during instance creation, then we can be 100% sure that the class will never be instantiated without its dependencies injected.
+
+The **IoC container makes sure all the arguments provided in the constructor are available before bean creation**. This helps in preventing the infamous `NullPointerException`.
+```java
+@Component
+public class Cake {
+
+	private static Logger LOGGER = LoggerFactory.getLogger(Cake.class);
+
+	private Flavor flavor;
+
+	public Cake(Flavor flavor) throws IllegalAccessException {
+		
+		//check if the required dependency is not null
+		if (flavor != null) {
+			this.flavor = flavor;
+			LOGGER.info("Flavor from Constructor Injection : " + flavor);
+		} else {
+			throw new IllegalArgumentException("Cake cannot be created with null flavor object");
+		}
+	}
+	
+	...
+}
+```
+
+In setter injection, these dependencies are initialized using setter methods only after the constructor invocation. Using setter injection, we can override the existing values of the object and provide different values. This kind of injection is preferable when we have optional dependencies.
+
+Constructor injection is extremely useful since we do not have to write separate business logic everywhere to check if the all the required dependencies are loaded thus simplifying code complexity.
+
+## Identifying Code smells
+
+Constructor injection can help us identify if our bean is dependent many other classes. If our constructor has large number of arguments, then we can be sure that our class has too many responsibilities. It is not a sign of quality code implying it should be refactored to better address proper separation of concerns.
+
+Constructor injection can help us find out circular dependencies in the code. If it exists, Spring would throw `BeanCurrentlyInCreationException` . This can help the developer identify classes that have circular dependency.
```

Comment:
I didn't know that! :) Are you sure we don't get an exception for circular dependencies with setter injection? Did you try that out?

Reply:
Yeah I checked and I got the same exception for setter injection. So I removed this point.